tc/q_cbq.c

```diff
@@ -63,7 +63,12 @@ static int cbq_parse_opt(struct qdisc_util *qu, int argc, char **argv, struct nl
 		if (matches(*argv, "bandwidth") == 0 ||
 		    matches(*argv, "rate") == 0) {
 			NEXT_ARG();
-			if (get_rate(&r.rate, *argv)) {
+			if (strchr(*argv, '%')) {
```

Comment:
We still have the issue that the user may have specified a "rate" option for one of the individual qdiscs but did not specify what the device is as part of the main `tc` command. So it's possible that we're reaching this case but `dev` doesn't contain a device name.

You need to check that the device name has been specified. For example, in `tc_class_modify()`, they do the following check (the char array that could hold the device name is called `d` in that function):

    if (d[0])  {
        ...

We need to make sure we check this for all functions that try to calculate a percentage. So maybe this check should be the first thing done in `parse_percent_rate()`?

Reply:
I think`tc_class_modify()` or `tc_qdisc_modify()` throw an error when no device is being specified. And the control goes from this code, and then to the qdisc file (which parses the rate, class arguments, etc) so doesn't that already take care of the 'no device specification scenario' ? (just thinking out loud here)

Reply:
But even in that case, I guess this would be a safety check inside `parse_percent_rate()` which is necessary. But do let me know what you think. Thanks.